### PR TITLE
Sharing: ask search engines to not index pages with query string.

### LINKED ISSUE #21836: Jetpack sharing buttons causing Google Search Console warnings
### Impacted plugin

Jetpack

### Steps to Reproduce

1. Enable sharing buttons
2. Set up Google Search Console
3. Google Search Console attempts to index the URLs linked on each sharing button (which is the post URL appended with ex.  `?share=facebook`, `?share=reddit`, and `?share=email`)

### A clear and concise description of what you expected to happen.

I wouldn't expect GSC to attempt to index these URLs, as they all link to outside the site (for example, the URL with `?share=facebook` opens a Facebook share panel from Facebook).

### What actually happened

GSC attempts to index the URLs, and it reports a warning: `indexed, though blocked by robots.txt`. This results in users receiving a warning/notice email, and they don't seem to be able to dismiss the warnings. 

### Other information

Opening an issue here as we've received at least three separate user reports about this over the past month or so.

Looking at the source code for the sharing buttons, I can see these URLs are added to the buttons with `rel="nofollow noopener noreferrer"` 

Have there been any changes made to this recently? Is this an issue on the GSC side, or is there a way we can change the way these URLs are linked to prevent the errors from occuring? 

### Operating System

_No response_

### OS Version

_No response_

### Browser

Chrome/Chromium

### Browser Version(s)

_No response_

### PR DESCRIPTION
**Pull request by jeherve**
Fixes #21836

#### Changes proposed in this Pull Request:

Jetpack's unofficial (icon, icon+text, text) sharing buttons currently work like so:
- When you append a query string with the name of a sharing service to a post URL, we redirect to the social service in question, with information the service can use (e.g. the post title as a tweet message).
- That query string may look like `?share=twitter`
- A page with such query string is not directly accessible since you are immediately redirected to the social network.
- WP Core implements `canonical` rel links in the head of each post page, so if you were looking at such a page's source code before the redirect happens, you would see that the canonical link for that page is in fact without the query string.
- Our HTML markup for the sharing buttons includes a `nofollow` rel attribute in the links, so as to indicate to search engines that the link should not be followed.

Yet that does not seem to be enough, as Google Search Console still seems to complain about those links in some scenarios, as mentioned in the issue above.

#### Jetpack product discussion

* N/A

#### Does this pull request change what data or activity we track or use?

* No

#### Testing instructions:

* Go to Settings > Reading and ensure your site is publicly accessible and indexable.
* Activate the Sharing module on your site.
* In Settings > Sharing, ensure the Twitter button is added.
* Load a published post or a page with sharing buttons.
* Click on the Twitter button
* Ensure that you are properly redirected, that you can share the post.
* Back to the post, copy the URL of the button instead of clicking on it.
* In your terminal, run `curl -I the-url-you-copied` and ensure that the response includes the following: `X-Robots-Tag: noindex, nofollow`
* Try with a bunch of other services, not just Twitter.

### CODE REVIEW

**Comment by anomiex** [comment]
I don't think the suggested testing instructions make sense. We want to look at what Google would actually see, not some random page WordPress outputs when you skip the actual redirect.

If I skip that and just hit the share URL with curl, this PR doesn't seem to do anything. The redirect comes with a zero-length document (so no `<meta>` tags) and there's no `X-Robots-Tag` header either.

See https://github.com/Automattic/jetpack/issues/21836#issuecomment-975820536 for more analysis. TL;DR: We probably want all the share redirects to do `header( 'X-Robots-Tag: noindex' )` instead.

**Comment by jeherve** [comment]
@anomiex Thanks for all the details. I've updated my PR accordingly. Let me know what you think.

File: projects/plugins/jetpack/modules/sharedaddy/sharing-sources.php [hunk 1]
```diff
@@ -911,9 +924,7 @@ public function process_request( $post, array $post_data ) {
 			'https://twitter.com/intent/tweet'
 		);
 
-		// Redirect to Twitter
-		wp_redirect( $twitter_url );
-		die();
+		parent::redirect_request( $twitter_url );
```
Comment (anomiex) [inline projects/plugins/jetpack/modules/sharedaddy/sharing-sources.php:927]: The use of `parent::` rather than `$this->` seems weird to me, but it works.

**Review by anomiex — APPROVED** [review]
PR looks ok now, but now I'm not entirely sure it'll fix the bug after all. When I try it now it seems like Google is checking the target's robots.txt before the redirect's `X-Robots-Tag`. But this is the best we can do on our end, if this doesn't do it then it'll have to be Google that fixes it on theirs.

File: projects/plugins/jetpack/modules/sharedaddy/sharing-sources.php (hunk 1 above)
Comment (jeherve) [reply projects/plugins/jetpack/modules/sharedaddy/sharing-sources.php:927]: I mostly followed the pattern in place for that class on this one.

**Pull request merged**